File: aif/storage.py
"""Attachment blobs on disk.

The database stores only metadata plus a generated uuid (``key``); the bytes live in
``cfg.attachments_dir/<key[:2]>/<key>`` so the whole folder can sit on a volume.
"""

from __future__ import annotations

import hashlib
import os
import re
import shutil
import uuid
from typing import BinaryIO

from . import sanitize
from .config import Config

CHUNK = 256 * 1024
_EXT_RE = re.compile(r"[^a-z0-9]")


class StorageError(RuntimeError):
    pass


class TooLarge(StorageError):
    pass


def sanitize_name(name: object, limit: int = 200) -> str:
    """Keep an uploaded name displayable: basename, no control/invisible chars, bounded length.

    Display only - blobs are always named by a generated uuid, never by client text.
    """
    raw = sanitize.fold(name).replace("\\", "/").rsplit("/", 1)[-1]
    return " ".join(raw.split())[:limit] or "file"


def sanitize_ext(name: str, limit: int = 8) -> str:
    ext = os.path.splitext(name)[1].lstrip(".").lower()[:limit]
    ext = _EXT_RE.sub("", ext)
    return f".{ext}" if ext else ""


def new_key() -> str:
    return uuid.uuid4().hex


def blob_path(cfg: Config, key: str) -> str:
    """Absolute path of a blob; rejects anything that is not a generated key."""
    if not re.fullmatch(r"[0-9a-f]{32}", key):
        raise StorageError(f"invalid blob key: {key!r}")
    return os.path.join(cfg.attachments_dir, key[:2], key)


def save(cfg: Config, stream: BinaryIO, key: str, limit: int) -> tuple[int, str]:
    """Stream ``stream`` into the blob store. Returns ``(size, sha256)``.

    Raises :class:`TooLarge` (and removes the partial blob) when ``limit`` is exceeded.
    """
    path = blob_path(cfg, key)
    os.makedirs(os.path.dirname(path), exist_ok=True)
    digest = hashlib.sha256()
    size = 0
    try:
        with open(path, "wb") as fh:
            while True:
                block = stream.read(CHUNK)
                if not block:
                    break
                size += len(block)
                if size > limit:
                    raise TooLarge(f"file exceeds limit of {limit} bytes")
                digest.update(block)
                fh.write(block)
    except BaseException:
        shutil.rmtree(os.path.dirname(path), ignore_errors=True) if not os.path.isdir(path) else None
        os.remove(path) if os.path.exists(path) else None
        raise
    return size, digest.hexdigest()


def open_blob(cfg: Config, key: str) -> BinaryIO:
    path = blob_path(cfg, key)
    if not os.path.isfile(path):
        raise StorageError(f"blob missing for key {key}")
    return open(path, "rb")


def remove(cfg: Config, key: str) -> bool:
    """Delete a blob, tolerating an already missing file."""
    path = blob_path(cfg, key)
    try:
        os.remove(path)
        return True
    except FileNotFoundError:
        return False
    except OSError as exc:  # pragma: no cover - depends on host FS
        raise StorageError(f"cannot delete blob {key}: {exc}") from exc


def stats(cfg: Config) -> dict[str, int]:
    total = 0
    count = 0
    for dirpath, _, files in os.walk(cfg.attachments_dir):
        for name in files:
            try:
                total += os.path.getsize(os.path.join(dirpath, name))
                count += 1
            except OSError:  # pragma: no cover - race with a purge
                continue
    return {"blobs": count, "bytes": total}

File: internal/storage/storage.go
package storage

import (
	"crypto/rand"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"github.com/dshein-alt/aif/internal/config"
	"github.com/dshein-alt/aif/internal/sanitize"
)

const chunkSize = 256 * 1024

var extRe = regexp.MustCompile(`[^a-z0-9]`)
var keyRe = regexp.MustCompile(`^[0-9a-f]{32}$`)

var ErrStorage = errors.New("storage error")
var ErrTooLarge = errors.New("file exceeds size limit")

// SanitizeName keeps an uploaded name displayable (display only; blobs are always named by uuid).
func SanitizeName(v any, limit int) string {
	raw := sanitize.Fold(v)
	raw = strings.ReplaceAll(raw, "\\", "/")
	if i := strings.LastIndex(raw, "/"); i >= 0 {
		raw = raw[i+1:]
	}
	raw = strings.Join(strings.Fields(raw), " ")
	if raw == "" {
		return "file"
	}
	return sanitize.Cap(raw, limit)
}

func NewKey() string {
	b := make([]byte, 16)
	rand.Read(b)
	return hex.EncodeToString(b)
}

func blobPath(cfg *config.Config, key string) (string, error) {
	if !keyRe.MatchString(key) {
		return "", ErrStorage
	}
	return filepath.Join(cfg.AttachmentsDir, key[:2], key), nil
}

// Save streams r into the blob store, returning (size, sha256hex). Removes partials on error.
func Save(cfg *config.Config, r io.Reader, key string, limit int64) (int64, string, error) {
	path, err := blobPath(cfg, key)
	if err != nil {
		return 0, "", err
	}
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return 0, "", err
	}
	f, err := os.Create(path)
	if err != nil {
		return 0, "", err
	}
	digest := sha256.New()
	var size int64
	buf := make([]byte, chunkSize)
	for {
		n, rerr := r.Read(buf)
		if n > 0 {
			size += int64(n)
			if size > limit {
				f.Close()
				os.Remove(path)
				return 0, "", ErrTooLarge
			}
			digest.Write(buf[:n])
			if _, werr := f.Write(buf[:n]); werr != nil {
				f.Close()
				os.Remove(path)
				return 0, "", werr
			}
		}
		if rerr == io.EOF {
			break
		}
		if rerr != nil {
			f.Close()
			os.Remove(path)
			return 0, "", rerr
		}
	}
	if err := f.Close(); err != nil {
		os.Remove(path)
		return 0, "", err
	}
	return size, hex.EncodeToString(digest.Sum(nil)), nil
}

// Open returns a reader for a stored blob.
func Open(cfg *config.Config, key string) (*os.File, error) {
	path, err := blobPath(cfg, key)
	if err != nil {
		return nil, err
	}
	f, err := os.Open(path)
	if errors.Is(err, os.ErrNotExist) {
		return nil, ErrStorage
	}
	return f, err
}

// ReadAll returns the full bytes of a stored blob.
func ReadAll(cfg *config.Config, key string) ([]byte, error) {
	f, err := Open(cfg, key)
	if err != nil {
		return nil, err
	}
	defer f.Close()
	return io.ReadAll(f)
}

func Remove(cfg *config.Config, key string) bool {
	path, err := blobPath(cfg, key)
	if err != nil {
		return false
	}
	if err := os.Remove(path); err != nil {
		return false
	}
	return true
}

func Stats(cfg *config.Config) (int, int64) {
	count := 0
	var total int64
	filepath.WalkDir(cfg.AttachmentsDir, func(_ string, d os.DirEntry, err error) error {
		if err != nil || d.IsDir() {
			return nil
		}
		if info, ierr := d.Info(); ierr == nil {
			total += info.Size()
			count++
		}
		return nil
	})
	return count, total
}
